text/0074/0074-entity-slicing.md

```diff
@@ -0,0 +1,566 @@
+
+```
+pub struct RequestType {
+    pub principal: EntityType,
+    pub action: EntityUID,
+    pub resource: EntityType,
+}
+```
+
+
+The `RequestEntityManifest` is a trie storing all the access paths required for the given request type. Access paths all start with Cedar variables and entity literals:
+
+```
+/// The root of an entity slice.
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Hash)]
+pub enum EntityRoot {
+    /// Literal entity ids
+    Literal(EntityUID),
+    /// A Cedar variable
+    Var(Var),
+}
+
+
+/// a [`RequestEntityManifest`] is a trie that specifies what data to load
+#[serde_as]
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct RequestEntityManifest {
+    #[serde_as(as = "Vec<(_, _)>")]
+    /// The data that needs to be loaded, organized by root
+    pub trie: HashMap<EntityRoot, AccessTrie>,
+}
+```
+
+
+Finally, we have the `AccessTrie` data structure, which stores the access paths trie. These paths can also request ancestors in the entity hierarchy. Note: in the implementation, the entity slice also stores optional type annotations.
+
+```
+pub struct AccessTrie {
+    /// Child data of this entity slice.
+    pub children: HashMap<SmolStr, Box<AccessTrie>>,
+    /// If the ancestors in the entity hierarchy are required.
+    pub ancestors_required: bool
+}
+```
+
+
+
+
+## Computing the Entity Manifest: High Level
+
+Computing the entity manifest has two parts: **calculating access paths** and **merging** **access paths.**
+For calculating access paths, we write a static analysis that finds all the relevant access paths in the policies. Cedar has only four variables (`principal`, `resource`, `action`, and  `context` ), so all data is loaded starting with a variable or entity literal.
+For merging access paths, we define a method for merging access paths, sharing common prefixes in a Trie.
+
+The high-level algorithm looks like this:
+
+1. First, perform Cedar type-checking to get type-annotated policies that are specific to the type of the request. 
+2. For each type of request
+    1. For each policy with that request type
+        1. **Compute access paths**
+    2. **Merge** all of the access paths into a single entity slice trie
+
+
+The merge operator for access paths is straightforward. First, convert all access paths directly to a Trie, where the fields in the path are the edges in the Trie. Then, use the standard merge operator on Tries.
+
+
+## Computing access paths
+
+This section proposes a simple static analysis that soundly computes all of the necessary access paths for a Cedar expression.
+While it’s possible to soundly handle all Cedar expressions, we simplify the problem by rejecting Cedar programs unless they follow the following grammar:
```

Comment:
I imagine we'll want to have a validation check that ensures that the static analysis will work. In terms of implementation, it should be straightforward to add an additional `ValidationMode` behind the `entity-manifest` flag.

Reply:
This is a similar idea to what I proposed in [an earlier thread](https://github.com/cedar-policy/rfcs/pull/74#discussion_r1700100194), but I go even further and propose that if we convince ourselves these restrictions are reasonable, we should just incorporate them into the existing `ValidationMode::Strict`

Reply:
Luckily, I've written a static analysis that works on all cedar programs making this unnecessary. I've updated the RFC.